### ISSUE TITLE
IDE keeps getting stuck in a loop to restart R when trying to install a package that is already installed

### ISSUE THREAD

**Issue by shalutiwari** [rstudio/rstudio#2656]
Was trying to install datasets package for an app and din't know if it was already installed at the startup. 

It kept showing a dialog box to install R and restart the session. Even after saying yes once it doesn't stop there and continues to do so until you hit cancel and then it will throw the error message below.

install.packages("datasets")
Error in install.packages : Updating loaded packages

This behavior is not intuitive and doesn't tell the user anything. Instead it can just check if the package exists and say its installed or doesn't need an upgrade.

<img width="606" alt="screen shot 2018-04-17 at 12 58 54 pm" src="https://user-images.githubusercontent.com/32690638/38884967-4e4855a8-423f-11e8-97c2-7afa424d9bb8.png">

**Comment by jmcphers** [comment]
This particular package is tricky because it's loaded at startup. RStudio presumes that the package will be unloaded when R is restarted, but it isn't. 

One possible solution would be to record the set of packages which were loaded at R startup, via e.g. an initial call to `loadedNamespaces()` (this would record base package names as well as any which were injected in `.Rprofile` and friends). If the user attempts to install one of these special packages, we could show a similar warning, but one that doesn't suggest or initiate a restart (just a "install anyway? y/n" prompt) since one isn't likely to be effective.

**Comment by aronatkins** [comment]
I have seen this issue with "plumber" and maybe a few other packages.

**Comment by aronatkins** [comment]
Here is how I was able to get into the restart loop with "plumber":

```
# restart R and clear environment before starting (to remove "p")
devtools::install_version("plumber", version = "0.4.2")
library(plumber)
p <- plumber::plumb(dir = ".")
install.packages("plumber")
```

If I don't have "p" in the environment, I am prompted about a restart but there is no loop. If I do remove the `library` statement and do not define "p", there is no prompt for a restart of R.

**Comment by aronatkins** [comment]
I was seeing the described behavior with "plumber" using RStudio version 1.2.605 and R 3.4.3.

**Comment by rstudio-build** [comment]
That happens whenever an object of a type that requires a namespace load is
sitting in the global environment. There's no workaround other than for the
user to just say "No". One possible remediation would be to limit this
behavior to only packages that have native code on Windows (which is where
the problem really is). I believe that this would solve the problem for
plumber.

**Comment by kevinushey** [comment]
Another possibility would be to provide a user gesture for restarting without restoring the global environment on restart, and providing that as an option in the dialog.

**Comment by R-Hannibal** [comment]
Ran into this issue twice today, once when installing the tidyverse package and again when updating the tibble package. 

This seemed particularly clunky with the tibble package update as I already have the package installed and am just trying to update it through the mechanisms provided within the IDE, and it yields an endless loop (*endless until the user gets frustrated and selects 'No' or 'Cancel'. The package still had not updated, had to remove and re-install from CRAN to get the correct version).

**Comment by ronblum** [comment]
Two recent duplicates:
- #8710 
- #8799

**Comment by david-cortes** [comment]
Not sure about windows, but in other OSes, wouldn't it make more sense to ask the user to restart **after** the package gets installed? i.e. "would you like to restart the session after the package is successfully installed".

**Comment by mirh** [comment]
This also happens whenever you try to reinstall bookdown with a bookdown project open.

**Comment by RailingResi** [comment]
I just had the same problem and could solve it with managing the packages via the 'Packages' tab in RStudio. 
<img width="1226" alt="Screenshot 2021-05-28 at 14 10 18" src="https://user-images.githubusercontent.com/22755523/119981890-70144480-bfbe-11eb-949a-fe5063e8f008.png">
Hope this helps anyone too.

**Comment by 7kemZmani** [comment]
As @jmcphers noted, I added packages in `.Rprofile` for convenience, so that they're loaded automaticlly in every new R session; yet they were causing the issue you're describing. Simply, disable (comment) those packages in order for Rstudio to proceed with the updates. 

<img width="1564" alt="Screenshot 2021-06-11 133734" src="https://user-images.githubusercontent.com/433779/121675427-0e72d080-cabc-11eb-9bc7-2a49795491c3.png">

**Comment by zkamvar** [comment]
I've been experiencing this issue as well (and I am an ardent believer of EIKIFJB, so it's not because I have a stale R environment being loaded), but I think I know the issue on my end at the very least: **It's because of [custom Rmarkdown site generators](https://bookdown.org/yihui/rmarkdown/rmarkdown-site.html#custom-site-generators).**

## Context

I am currently beta testing a site generator (https://github.com/carpentries/sandpaper) with folks and am finding that this loop happens whenever someone tries to update this package inside the project in RStudio. I found this to be true of RStudio cloud projects as well. This has been observed on Ubuntu Linux 20.04 and MacOS

## Symptoms

If I inspect the `sessionInfo()` for different projects after loading them and reloading them with a clean slate, I notice that the context-specific engines are loaded:

### Example: blogdown

Packages needed for blogdown are loaded because of [the site directive in the index.Rmd file](https://github.com/zkamvar/zkamvar/blob/4cc973b11ed38000cd4d61a6e5b314a557615505/index.Rmd)

```r
usethis::create_from_github("zkamvar/zkamvar")
```

```
> sessionInfo()
R version 4.1.2 (2021-11-01)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] compiler_4.1.2  bookdown_0.24   fastmap_1.1.0   cli_3.2.0       htmltools_0.5.2
 [6] tools_4.1.2     yaml_2.3.4      rmarkdown_2.11  blogdown_1.7    knitr_1.37     
[11] xfun_0.29       digest_0.6.29   rlang_1.0.1     evaluate_0.14  
```

### Example: sandpaper

Packages needed for sandpaper are loaded because of [the site directive in the index.md file](https://github.com/carpentries/sandpaper-docs/blob/978bf9f3800e01bf79e8d11abc567b13ce3c9529/index.md)

```r
usethis::create_from_github("zkamvar/zkamvar")
```

```
> sessionInfo()
R version 4.1.2 (2021-11-01)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] ps_1.6.0         rprojroot_2.0.2  digest_0.6.29    assertthat_0.2.1
 [5] R6_2.5.1         evaluate_0.14    rlang_1.0.1      cli_3.2.0       
 [9] fs_1.5.2         sandpaper_0.1.6  callr_3.7.0      rmarkdown_2.11  
[13] tools_4.1.2      xfun_0.29        yaml_2.3.4       fastmap_1.1.0   
[17] compiler_4.1.2   processx_3.5.2   htmltools_0.5.2  knitr_1.37  
```

**Comment by kathleensuvada2022** [comment]
Hi- I also just had this issue and it seems my R itself was outdated and my R studio was up to date. Updating R itself seems to have solved the looping issue for me.

**Comment by hissarious** [comment]
> Hi- I also just had this issue and it seems my R itself was outdated and my R studio was up to date. Updating R itself seems to have solved the looping issue for me.

This worked for me as well

**Comment by iyliahaziq11** [comment]
not sure if this would help anyone. i was facing the same issue, with R Studio prompting me the same message and getting stuck in a loop.

i did a use_python in R Studio, pointing to the python in anaconda, after which I could magically run nevergrad. 

**use_python("/Users/myMac/opt/anaconda3/bin/python")**

**Comment by mikebessuille** [comment]
I had this happen to me today...

**Comment by trinetra75** [comment]
I have a similar symptom: 
restarting a session a development package is attempted to load via the `library()` command
being a development package this fails but `rstudio` continues to try to load the package until
I manually load it via `devtools`...

Example of output:

```
Loading required package: COTAN
Loading required package: COTAN
> devtools::load_all("~/dev/COTAN/COTAN/")
ℹ Loading COTAN
There were 50 or more warnings (use warnings() to see the first 50)
```

Should I submit a separate issue?

**Comment by plorch** [comment]
Many of the examples here are special cases of the general problem of updating a package when Rstudio (or R in some cases) has loaded certain package on startup.  My solution is to keep an Rstudio project in some location like my Documents folder where I never save the workspace or history when I quit.  I can then open this and update packages.  In rare cases, I still need to delete a package from the R library and install.package() it.

**Comment by chrisoswald** [comment]
I have the same problem and found that the reason was "rstudioapi". After installing rstudioapi manual, the rest was working

**Comment by eragrostis** [comment]
This issue continues. I wasn't able to update rlang or several other packages in rstudio and had to load just R and update.packages()

**Comment by kevinushey** [comment]
For testing, create a Bookdown project, and then try running:

```
install.packages("rmarkdown")
```

Bookdown projects require `rmarkdown`, and so the package is loaded somewhat eagerly. The PR in https://github.com/rstudio/rstudio/pull/14413 should ensure that package installation happens before the `rmarkdown` package is loaded, and so should succeed.

**Comment by ronblum** [comment]
Verified in RStudio Desktop 2024.04.0-daily+573 on MacOS 14.4. Tried a number of things, including:
- Plumber: https://github.com/rstudio/rstudio/issues/2656#issuecomment-386717406
- Bookdown: https://github.com/rstudio/rstudio/issues/2656#issuecomment-1992662337
- `library(dplyr)` followed by `install.packages("dplyr")`

Choosing "Yes" will only restart R once, and then installation is successful—no more looping. Choosing "No" or "Cancel" behave as before.
